examples/simple/src/lib.rs

```diff
@@ -9,11 +9,21 @@ use numpy::{
 use pyo3::{
     pymodule,
     types::{PyDict, PyModule},
-    FromPyObject, PyAny, PyResult, Python,
+    exceptions::PyIndexError,
+    FromPyObject, PyAny, PyObject, PyResult, Python,
 };
 
 #[pymodule]
 fn rust_ext(_py: Python<'_>, m: &PyModule) -> PyResult<()> {
+    // example using generic PyObject
+    fn head(x: ArrayViewD<'_, PyObject>) -> PyResult<PyObject> {
+        if x.len() > 0 {
```

Comment:
Thank you for the PR!

Could you do

```rust
x.get(0).clone().ok_or_else(|| PyIndexError::new_err("array index out of range"))?
```

here instead of checking the length manually?

Reply:
Done! The expression you suggested gives the following compilation error

```
$ pipenv run maturin develop
🔗 Found pyo3 bindings with abi3 support for Python ≥ 3.7
🐍 Not using a specific python interpreter
   Compiling pyo3-build-config v0.16.5
   Compiling pyo3-ffi v0.16.5
   Compiling pyo3 v0.16.5
   Compiling numpy v0.16.2 (/Users/qys/Documents/Chuxiaof/tangram/rust-numpy)
   Compiling rust-ext v0.1.0 (/Users/qys/Documents/Chuxiaof/tangram/rust-numpy/examples/simple)
error[E0308]: `?` operator has incompatible types
  --> src/lib.rs:20:9
   |
20 |         x.get(0).clone().ok_or_else(|| PyIndexError::new_err("array index out of range"))?
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected enum `Result`, found `&Py<PyAny>`
   |
   = note: `?` operator cannot convert from `&Py<PyAny>` to `Result<Py<PyAny>, PyErr>`
   = note:   expected enum `Result<Py<PyAny>, PyErr>`
           found reference `&Py<PyAny>`


error: aborting due to previous error


   For more information about this error, try `rustc --explain E0308`.

error: could not compile `rust-ext` due to 2 previous errors
💥 maturin failed
  Caused by: Failed to build a native library through cargo
  Caused by: Cargo build finished with "exit status: 101": `"cargo" "rustc" "--message-format" "json" "--lib" "--" "-C" "link-arg=-undefined" "-C" "link-arg=dynamic_lookup" "-C" "link-args=-Wl,-install_name,@rpath/rust_ext.cpython-310-darwin.so"`
```

As a result, I used the following instead

```
x.get(0).and_then(|elem| Some(elem.clone())).ok_or_else(|| PyIndexError::new_err("array index out of range"))
```

I have also updated the test to make the NumPy array more "generic":

```
x = np.array(['first', None, 42])
```

Reply:
> Done! The expression you suggested gives the following compilation error

Sorry, there was a typo there: The method is `cloned` not `clone`, c.f. https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.cloned. Please try that instead of `and_then?.

Reply:
Done! The `cloned` version works after I wrap the whole thing in an `Ok`.

Reply:
Sorry, my mistake again: Just leave out both `Ok` and `?`, then it should work as well.

Reply:
You are right. Should have noticed that!